### PR TITLE
Migrate transducer input checks to C++

### PR DESCRIPTION
**Pull request by carolineechen**
Move input checks for RNNT from Python to C++

RNN Transducer Loss Issue: #1240

### CODE REVIEW

File: torchaudio/csrc/transducer.cpp [hunk 1]
```diff
@@ -17,11 +17,38 @@ int64_t cpu_rnnt_loss(
...
+  TORCH_CHECK(labels.is_contiguous(), "labels must be contiguous");
+  TORCH_CHECK(
+      label_lengths.is_contiguous(), "label_lengths must be contiguous");
+  TORCH_CHECK(input_lengths.is_contiguous(), "lengths must be contiguous");
+  TORCH_CHECK(
+      input_lengths.size(0) == acts.size(0), "must have a length per example.");
+  TORCH_CHECK(
+      label_lengths.size(0) == acts.size(0),
+      "must have a label length per example.");
+  TORCH_CHECK(acts.dim() == 4, "log_probs must be 4D");
```
Comment (mthrok) [inline torchaudio/csrc/transducer.cpp:36]: I am aware that these error messages are carried out from the original code, but I do not think `must be 4D` is user friendly. If the function is expecting the fixed shape, it should communicate that.  `4D (foo, bar, baz, boo)`.

File: torchaudio/csrc/transducer.cpp [hunk 2]
```diff
@@ -17,11 +17,38 @@ int64_t cpu_rnnt_loss(
...
+      "label_lengths must be int32 type");
+  TORCH_CHECK(
+      input_lengths.dtype() == torch::kInt32, "lengths must be int32 type");
+  TORCH_CHECK(acts.is_contiguous(), "log_probs must be contiguous");
+  TORCH_CHECK(labels.is_contiguous(), "labels must be contiguous");
+  TORCH_CHECK(
+      label_lengths.is_contiguous(), "label_lengths must be contiguous");
+  TORCH_CHECK(input_lengths.is_contiguous(), "lengths must be contiguous");
+  TORCH_CHECK(
+      input_lengths.size(0) == acts.size(0), "must have a length per example.");
```
Comment (mthrok) [inline torchaudio/csrc/transducer.cpp:32]: Again, I am aware that these error messages are carried out from the original code, but `"must have a length per example."` is not helpful on what to do. @carolineechen Can you look into the code and rephrase it to a good instruction?

**Review by mthrok — APPROVED** [review]

File: torchaudio/csrc/transducer.cpp [hunk 3]
```diff
@@ -17,11 +17,40 @@ int64_t cpu_rnnt_loss(
...
   int maxT = acts.size(1);
   int maxU = acts.size(2);
   int minibatch_size = acts.size(0);
   int alphabet_size = acts.size(3);
 
+  TORCH_CHECK(
+      at::max(input_lengths).item().toInt() == maxT, "input length mismatch");
+  TORCH_CHECK(
+      at::max(label_lengths).item().toInt() + 1 == maxU,
+      "output length mismatch");
```
Comment (vincentqb) [inline torchaudio/csrc/transducer.cpp:52]: (Follow-up beyond this PR: we'll want to improve this message too)

**Review by vincentqb — COMMENTED** [review]
Thanks for working on this! I've noted a few small changes (and some things we'll think about in separate PRs)
